### ISSUE TITLE
issue of smb2_stat/smb2_opendir on windows

### ISSUE THREAD

**Issue by hbzhang3** [sahlberg/libsmb2#137]
steps to reproduce
1. prepare two win10 systems A and B
2. share C:\users folder of A, 
3. run following codes on B

void test()
{
	smb2_stat_64 st = { 0 };
	smb2_context *smb2 = smb2_init_context();
	smb2_set_user(smb2, "username");
	smb2_set_password(smb2, "password");
	smb2_connect_share(smb2, "ip of A", "Users", NULL);  // succeed
	smb2_stat(smb2, "All Users", &st); // fail as expected because "All Users" folder is inaccessible
	smb2dir *d = smb2_opendir(smb2, "");  // fail, unexpected!
	if (d)
		smb2_closedir(smb2, d);
	smb2_destroy_context(smb2); // sometimes stack corruption crash happens here
}

**Comment by sahlberg** [comment]
Can not reproduce.

**Comment by timmeh87** [comment]
hello, I am on a windows 10 system, and using libsmb2 to connect to a centos system with a smb2 share. My program flow is somewhat complex and I dont have a minimum example, but I am sporradically getting a crash (very rare) from smb2_destroy_context(). The exception says "read access violation, fh was 0xfff...ff7"  and it is somewhere inside of "open_cb" when this happens.

I'm not sure if this is related or if I should raise a new issue or something, or if anyone will even read this

**Comment by sahlberg** [comment]
On Thu, May 14, 2020 at 4:18 PM timmeh87 <notifications@github.com> wrote:

> hello, I am on a windows 10 system, and using libsmb2 to connect to a
> centos system with a smb2 share. My program flow is somewhat complex and I
> dont have a minimum example, but I am sporradically getting a crash (very
> rare) from smb2_destroy_context(). The exception says "read access
> violation, fh was 0xfff...ff7" and it is somewhere inside of "open_cb" when
> this happens.
>
> I'm not sure if this is related or if I should raise a new issue or
> something, or if anyone will even read this
>

Can you build it with debugging , i.e. "-g -O0" and then next time you get
a core dump to load it into gdb and get a backtrace on where it crashed?

**Comment by timmeh87** [comment]
Hello, thank you for your reply. I did some more testing last night. I
updated libsmb2 to the latest commit (the bleeding edge, not 3.0.0) and
rebuilt it on the same machine Im using it on. I have a test program that
reads and writes many hundreds of files over smb, while it is doing other
things too. My network setup is that the SMB server is on spotty 2.4g WiFi
(residential cisco gateway), so every once in a while I get errors related
to network connectivity. This is on purpose, the software should be robust
enough to recover from such errors. My main issue is that the access
violation that comes from inside smb2_detroy_context is an "uncatchable"
exception so it is unrecoverable

I have observed that before the crash, there is usually (I hesitate to say
"always", but possibly always) a network disruption that
causes smb2_write() to return -1. At this point there is no crash and I can
continue doing things. Sometimes I will try to write a few other files, all
resulting in -1, and then I will try to destroy the context. At that point
it crashes.

my build environment is visual studio 2017, i am able to capture the exact
line the exception occurs on, and it has happened once since I updated but
I did not record the exact location. I will continue to do tests and return
here with more information.
